storage/SplitRowStoreTupleStorageSubBlock.cpp

```diff
@@ -194,379 +257,125 @@ TupleStorageSubBlock::InsertResult SplitRowStoreTupleStorageSubBlock::insertTupl
+
+            header_->variable_length_bytes_allocated += attr_size;
+          } else {  // nullable, variable length
+            DCHECK_EQ(static_cast<int>(current_null_idx), relation_.getNullableAttributeIndex(accessor_attr_id));
+
+            TypedValue attr_value(accessor->getTypedValue(attribute_map[accessor_attr_id]));
+            if (attr_value.isNull()) {
+              tuple_null_bitmap.setBit(current_null_idx, true);
             } else {
-              attr_value.copyInto(fixed_length_attr_storage
-                                  + relation_.getFixedLengthAttributeOffset(attr_it->getID()));
+              DCHECK_EQ(relation_.getVariableLengthAttributeIndex(accessor_attr_id),
+                        insertInfo.var_len_offsets_[accessor_attr_id]);
+
+              const std::size_t attr_size = attr_value.getDataSize();
+              current_variable_position -= attr_size;
+              const int var_len_info_idx = insertInfo.var_len_offsets_[accessor_attr_id] * 2;
+              variable_length_info_array[var_len_info_idx] = current_variable_position;
+              variable_length_info_array[var_len_info_idx + 1] = attr_size;
+              attr_value.copyInto(static_cast<char *>(tuple_storage_) + current_variable_position);
+              header_->variable_length_bytes_allocated += attr_size;
             }
-          }
-          occupancy_bitmap_->setBit(pos, true);
-          ++(header_->num_tuples);
-          if (pos > header_->max_tid) {
-            header_->max_tid = pos;
+            current_null_idx++;
           }
         }
-      } else {
-        while (accessor->next()) {
-          pos = this->isPacked() ? header_->num_tuples
-                                 : occupancy_bitmap_->firstZero(pos);
-          if (!this->spaceToInsert(pos, 0)) {
-            accessor->previous();
-            break;
-          }
-          void *tuple_slot = static_cast<char*>(tuple_storage_) + pos * tuple_slot_bytes_;
-          char *fixed_length_attr_storage = static_cast<char*>(tuple_slot) + per_tuple_null_bitmap_bytes_;
-
-          std::vector<attribute_id>::const_iterator attr_map_it = attribute_map.begin();
-          for (CatalogRelationSchema::const_iterator attr_it = relation_.begin();
-               attr_it != relation_.end();
-               ++attr_it, ++attr_map_it) {
-            const void *attr_value = accessor->template getUntypedValue<false>(*attr_map_it);
-            std::memcpy(fixed_length_attr_storage
-                            + relation_.getFixedLengthAttributeOffset(attr_it->getID()),
-                        attr_value,
-                        attr_it->getType().maximumByteLength());
-          }
-          occupancy_bitmap_->setBit(pos, true);
-          ++(header_->num_tuples);
-          if (pos > header_->max_tid) {
-            header_->max_tid = pos;
-          }
+        occupancy_bitmap_->setBit(pos, true);
```

Comment:
All of this would go away if we drop the occupancy_bitmap_. 

On the other hand, if we absolutely want to allow deletes, it'd be faster (and more space-efficient) to keep a sorted list of tuple IDs rather than an entire bitmap.